File: CSharp/125-ValidPalindrome/125ValidPalindrome.cs
// FOR LOOP
// Time Complexity: O(n) (Analyze Complexity Confirmed)
// You have to go through up to every letter in the string.
// Space Complexity: O(1)
// We aren't allocating any extra space, just traversing the string given.

public class Solution {
    public bool IsPalindrome(string s) {
        for (int i = 0, j = s.Length - 1; i < j;){
            if (!Char.IsLetterOrDigit(s[i])){i++; continue;}
            if (!Char.IsLetterOrDigit(s[j])){j--; continue;}
            if(Char.ToLower(s[i]) != Char.ToLower(s[j])){return false;}
            i++;
            j--;
        }
        return true;
    }
}

File: Javascript/125-ValidPalindrome/125ValidPalindrome.js
// There are two approaches for this. 
// Reverse the string
// Two pointers. 
// Obviously two pointers is longer but it's probably given as a solution because it's an easy introduction to two pointers.
// so I should come up with a javascript implementation for that to make sure I have it


/* REVERSE STRING
Time Complexity: O(n)
Must go through the entire string to clean it up and make it lower case.

Space Compexlity: O(n)
need to store the cleaned string in a new variable the same size as the initial string
*/


/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    const cleanStr = s.toLowerCase().replace(/[^a-z0-9]/g, "");
    return cleanStr === cleanStr.split("").reverse().join("");
};
